internal/disabler/disabler.go

```diff
@@ -0,0 +1,111 @@
+// package disabler contains a component that can be enabled and disabled dynamically
+package disabler
+
+import (
+	"context"
+	"errors"
+
+	"github.com/pomerium/pomerium/internal/log"
+)
+
+var errDisabled = errors.New("disabled")
+
+type Handler interface {
+	RunEnabled(ctx context.Context) error
+}
+
+type Disabler interface {
+	Run(ctx context.Context) error
+	Enable()
+	Disable()
+}
+
+type disabler struct {
+	name            string
+	handler         Handler
+	onChangeEnabled chan bool
+}
+
+// New creates a new Disabler. When the Disabler is enabled, the Handler's
+// RunEnabled will be called. If the Disabler is subsequently disabled the
+// context passed to RunEnabled will be canceled. If the Disabler is subseqently
+// enabled, RunEnabled will be started again.
+func New(name string, handler Handler, enabled bool) Disabler {
+	d := disabler{name: name, handler: handler, onChangeEnabled: make(chan bool, 1)}
+	d.change(enabled)
+	return d
+}
+
+func (d disabler) Run(ctx context.Context) error {
+	for {
+		// listen for a transition to enabled
+		var enabled bool
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case enabled = <-d.onChangeEnabled:
+		}
+		if !enabled {
+			// wait until we're enabled
+			continue
+		}
+
+		log.Ctx(ctx).Info().Msgf("enabled %s", d.name)
+		err := d.runEnabledOnce(ctx)
+		if errors.Is(err, errDisabled) {
+			log.Ctx(ctx).Info().Msgf("disabled %s", d.name)
+			continue
+		}
+
+		// for any non-"disabled" error, we return it
+		return err
+	}
+}
+
+func (d disabler) runEnabledOnce(ctx context.Context) error {
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(context.Canceled) // ensure we cancel the context if we exit early
+
+	// start a background goroutine that will listen for a transition to disabled
+	go func() {
```

Comment:
I wonder if we do need a background goroutine for this? I haven't completely thought it through yet but I wonder if we could simplify this a little by storing the `cancel` function in a `sync.Pointer` field on the `disabler` struct.

I'm thinking the Enable operation would create a new `context.WithCancelCause(ctx)`, then `CompareAndSwap(nil, cancel)`, and then call `RunEnabled()` if the swap succeeded. The Disable operation could retrieve the cancel function using `Swap(nil)` and then invoke it if it isn't nil.

Does that make any sense? Any thoughts / feelings?

Reply:
Run happens in a separate goroutine from Enable/Disable. I'm not sure how to communicate the state change apart from using a channel.

We could wait for `enableCtx.Done()` and cancel it in Enable and then wait for `disableCtx.Done()` and cancel it in Disable... but I think that wouldn't work and fixing it feels racy.

I suppose if we track the current state and wrap that with a mutex we could avoid the race. I can try to implement this.

Reply:
Why do we want to avoid using a goroutine?

Reply:
Using a mutex did simplify the code.

Reply:
OK great, like I said I probably didn't think it through all the way. I think I didn't really understand the contract for the `Run()` method. I'll defer to your judgement on whatever looks simplest.